moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/ompl_optimization_objective_loader.h

```diff
@@ -0,0 +1,75 @@
+ *     with the distribution.
+ *   * Neither the name of Willow Garage nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Antoine Duplex */
+
+#ifndef SRC_OMPL_OPTIMIZATION_OBJECTIVE_LOADER_H
+#define SRC_OMPL_OPTIMIZATION_OBJECTIVE_LOADER_H
+
+#include <ompl/base/OptimizationObjective.h>
+#include <ompl/base/SpaceInformation.h>
+
+#include <string>
+#include <iostream>
+
+#define MOVEIT_OPTIMIZATION_OBJECTIVE_PLUGIN(my_namespace, OptimizationObjective)                                      \
+  namespace my_namespace                                                                                               \
+  {                                                                                                                    \
+  class OptimizationObjective##Loader : public ompl_optimization_loader::OptimizationObjectiveLoader                   \
+  {                                                                                                                    \
+  public:                                                                                                              \
+    ompl::base::OptimizationObjectivePtr getOptimizationObjective(ompl::base::SpaceInformationPtr si) override         \
+    {                                                                                                                  \
+      return std::make_shared<OptimizationObjective>(si);                                                              \
+    }                                                                                                                  \
+  };                                                                                                                   \
+  PLUGINLIB_EXPORT_CLASS(my_namespace::OptimizationObjective##Loader,                                                  \
+                         ompl_optimization_loader::OptimizationObjectiveLoader)                                        \
+  }
```

Comment:
Yes indeed it would have been easier, unfortunately the constructor of the class OptimizationObjective needs an argument, the SpaceInformation, and PluginLib doesn't allow to load such class. That is why we need a loader that can be instantiated without any argument and then load the optimization objective.

Reply:
The macro makes it easier to the user. He just has to call at the end of his new OMPL's optimization objective class to inherit and instantiate the loader.